src/lib/EDSL.ml

```diff
@@ -111,3 +109,174 @@ let if_seq ~t ?e c =
   match e with
   | None -> if_then c (seq t)
   | Some f -> if_then_else c (seq t) (seq f) 
+
+
+let eprintf fmt l =
+  with_redirections
+    (call (string "printf" :: string "--" :: fmt :: l)) [
+    to_fd (int 1) (int 2);
+  ]
+
+module Command_line = struct
+  type 'a cli_option = {
+    switches: string list;
+    doc: string;
+    default: 'a;
+  }
+  type _ option_spec =
+    | Opt_flag:   bool t cli_option -> bool t option_spec
+    | Opt_string: string t cli_option -> string t option_spec
+  and (_, _) cli_options =
+    | Opt_end: string -> ('a, 'a) cli_options
+    | Opt_cons: 'c option_spec * ('a, 'b) cli_options -> ('c -> 'a, 'b) cli_options
+
+  module Arg = struct
+    let string ?(default = string "") ~doc switches =
+      Opt_string {switches; doc; default}
+    let flag ?(default = bool false) ~doc switches =
+      Opt_flag {switches; doc; default}
+
+    let (&) x y = Opt_cons (x, y)
+    let usage s = Opt_end s
+
+  end
+
+  let parse
+      (options: ('a, unit t) cli_options)
+      (action: anon: string list t -> 'a) : unit t =
+    let prefix = Common.Unique_name.variable "getopts" in
+    let variable {switches; doc;} =
+      sprintf "%s_%s" prefix (String.concat ~sep:"" switches|> Digest.string |> Digest.to_hex) in
+    let inits = ref [] in
+    let to_init s = inits := s :: !inits in
+    let cases = ref [] in
+    let to_case s = cases := s :: !cases in
+    let help_intro = ref "" in
+    let help = ref [] in
+    let to_help s = help := s :: !help in
+    let string_of_var var =
+      getenv (string var) in
+    let bool_of_var var =
+      getenv (string var) |> Bool.of_string in
+    let anonarg_var = prefix ^ "_anon" |> string in
+    let anon =
+      getenv anonarg_var |> string_list_of_string in
+    let unit_t =
```

Comment:
It's called `unit_t` because it's the result of applying the function passed by the user (`action`) to the all the arguments, but it could be better named indeed :)

For now, for the “how to use”, it's in the tests/examples: https://github.com/hammerlab/genspio/blob/master/src/test/examples.ml#L161
We should add more small examples though.

The function is building 3 pieces of code at once:

* variable intializations
* variable assignments in the `while true { switch { .... }` loop
* the `unit_t` : call the `action` function on the parsed arguments

The GADT typing of the loop that goes through the difflist (`type a b.  a -> (a, b) cli_options -> b`) is easier when we agglomerate the 2 first items in a `t list ref`.